### PR TITLE
feat(common): always clog GCP_LOG(FATAL) messages

### PR DESCRIPTION
**Pull request by devbww**
It is desirable that FATAL log messages are never discarded,
especially now that GCP_LOG(FATAL) calls `std::abort` (see #7058).

To that end, ...

- Don't allow FATAL logging to be compile-time disabled.

- Always instantiate a `StdClogBackend` in `DefaultLogBackend()`.

- Give that default `StdClogBackend` a "mininum log severity" from
  the first of the following conditions:
    * If `${GOOGLE_CLOUD_CPP_EXPERIMENTAL_LOG_CONFIG}` is valid,
      use the lowest enabled severity.
    * If `${GOOGLE_CLOUD_CPP_ENABLE_CLOG}` is a valid severity,
      use that.
    * If `${GOOGLE_CLOUD_CPP_ENABLE_CLOG}` is otherwise set,
      use the lowest enabled severity (backwards compatibility).
    * If `${GOOGLE_CLOUD_CPP_ENABLE_CLOG}` is unset, use FATAL.

This means that FATAL messages should always be logged. But note
that `LogSink::DisableStdClog()` and `LogSink::ClearBackends()`
can remove that `StdClogBackend`, so they should be used with care
(undoubtedly only from tests).

Also tidy up some tests so that the disposition of `LogBackend`s
is a little clearer.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7087)

### CODE REVIEW

File: google/cloud/internal/log_impl_test.cc [hunk 1]
```diff
@@ -68,62 +68,85 @@ TEST(DefaultLogBackend, CircularBuffer) {
...
 }
 
 TEST(DefaultLogBackend, CLog) {
   ScopedEnvironment config(kLogConfig, "clog");
   ScopedEnvironment clog(kEnableClog, absl::nullopt);
   auto be = DefaultLogBackend();
-  EXPECT_TRUE(!!be);
-  EXPECT_NE(dynamic_cast<StdClogBackend*>(be.get()), nullptr);
+  auto const* clog_be = dynamic_cast<StdClogBackend*>(be.get());
+  ASSERT_NE(clog_be, nullptr);
```
Comment (coryan) [inline google/cloud/internal/log_impl_test.cc:85]: nit: here and below, consider `ASSERT_THAT(clog_be, IsNotNull());`

Comment (coryan) [reply google/cloud/internal/log_impl_test.cc:85]: And the one above too.

File: google/cloud/internal/log_impl.cc [hunk 2]
```diff
@@ -22,9 +22,11 @@ namespace internal {
 
 void StdClogBackend::Process(LogRecord const& lr) {
   std::lock_guard<std::mutex> lk(mu_);
-  std::clog << lr << "\n";
-  if (lr.severity >= Severity::GCP_LS_WARNING) {
-    std::clog << std::flush;
+  if (lr.severity >= min_severity_) {
```
Comment (coryan) [inline google/cloud/internal/log_impl.cc:25]: optional: `if (lr.severity < min_severity_) return;` reduces indentation.

**Review by coryan — APPROVED** [review]
Thanks for this change.  A couple of minor nits, completely your choice.

File: google/cloud/internal/log_impl_test.cc (hunk 1 above)
Comment (devbww) [reply google/cloud/internal/log_impl_test.cc]: Done.

File: google/cloud/internal/log_impl.cc (hunk 2 above)
Comment (devbww) [reply google/cloud/internal/log_impl.cc:25]: I often find the single-entry/exit and indentation (within reason) clearer, and this is such an example, so I left it alone.

**Pull request merged**